local-cluster/src/cluster_tests.rs

```diff
@@ -527,20 +532,39 @@ pub fn start_gossip_voter(
                 });
 
                 for (parsed_vote, leader_vote_tx) in &parsed_vote_iter {
-                    if let Some(latest_vote_slot) = parsed_vote.last_voted_slot() {
+                    if let Some(vote_slot) = parsed_vote.last_voted_slot() {
+                        latest_vote_slot = vote_slot;
                         info!("received vote for {}", latest_vote_slot);
                         process_vote_tx(
                             latest_vote_slot,
                             leader_vote_tx,
                             parsed_vote,
-                            &cluster_info,
+                            &mut votes_to_sendout,
                         )
                     }
-                    // Give vote some time to propagate
-                    sleep(Duration::from_millis(sleep_ms));
                 }
 
-                if parsed_vote_iter.is_empty() {
+                // When push_active_set() is first called, Gossip doesn't have correct shred_version set, neither
+                // does it have any ping responses from peers, so messages pushed at this point will not be sent
+                // correctly. Wait until CRDS_GOSSIP_PULL_CRDS_TIMEOUT_MS / 2 + 1000 ms to give Gossip time to
+                // discover peers and set correct shred_version.
+                if timestamp() - start_time > (CRDS_GOSSIP_PULL_CRDS_TIMEOUT_MS / 2 + 1000)
+                    && !votes_to_sendout.is_empty()
+                {
+                    while let Some((vote, vote_index)) = votes_to_sendout.pop_front() {
+                        cluster_info.push_vote_at_index(vote, vote_index);
+                    }
+                    if max_vote_slot != latest_vote_slot {
+                        let new_epoch_slots: Vec<Slot> =
+                            (max_vote_slot + 1..latest_vote_slot + 1).collect();
+                        info!(
+                            "Simulating epoch slots from our node: {:?}",
+                            new_epoch_slots
+                        );
+                        cluster_info.push_epoch_slots(&new_epoch_slots);
+                        max_vote_slot = latest_vote_slot;
+                    }
+                    // Give vote some time to propagate
```

Comment:
hmm not quite sure I understand why did this change introduce this bug? Or is this one of those cases where the bug already existed but this change made it worse?

Reply:
ya since gossip proofs weren't being ingested we never got into the case 4b where the node stops voting and has to rely on epoch slots.